### PR TITLE
pdata: add [Type]Slice.Sort(func) to sort slices

### PR DESCRIPTION
**Pull request by jacobmarble**
**Description:**

@tigrannajaryan 
This change helps my efforts to validate the recent changes to the pdata package, [as requested](https://github.com/open-telemetry/opentelemetry-collector/issues/3104#issuecomment-871760928)

Comparing two generated slices can be tedious when the slices are not
ordered, and order does not matter for the test.

This change adds a .Sort(func(i, j int) bool) method to generated [Type]Slice. This is
similar to method StringMap.Sort(), but takes a less function because
the order may differ by context, where the sort order of a map is obvious (by key).

Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.>

**Documentation:**

Added some godoc.

### CODE REVIEW

**Comment by bogdandrutu** [comment]
@jacobmarble so this is mostly for testing correct? Need a bit of motivation and where this is needed.

**Comment by jacobmarble** [comment]
> @jacobmarble so this is mostly for testing correct? Need a bit of motivation and where this is needed.

Correct, this is *mostly* for testing.

I found one use of `StringMap.Sort()` outside of tests:
https://github.com/open-telemetry/opentelemetry-collector/blob/f9dacb8402fe2bd6d90a63b7f8ba084fd1bb1600/exporter/prometheusexporter/accumulator.go#L374-L384

I also found one use of `AttributeMap.Sort()` outside of tests:
https://github.com/open-telemetry/opentelemetry-collector/blob/c6f9f29fca3dce717275bda3263bf9c55f363823/internal/otlptext/databuffer.go#L303-L313

Neither of the above `Sort()` calls could be replaced by an `Equals(other)` method. Equivalence is what I really need; sorting `TypeSlice` objects allows me to do the following (I haven't found a better way to sort buckets yet):

```golang
func SortResourceMetrics(rm pdata.ResourceMetricsSlice) {
	for i := 0; i < rm.Len(); i++ {
		r := rm.At(i)
		for j := 0; j < r.InstrumentationLibraryMetrics().Len(); j++ {
			il := r.InstrumentationLibraryMetrics().At(j)
			for k := 0; k < il.Metrics().Len(); k++ {
				m := il.Metrics().At(k)
				switch m.DataType() {
				case pdata.MetricDataTypeGauge:
					for l := 0; l < m.Gauge().DataPoints().Len(); l++ {
						m.Gauge().DataPoints().At(l).LabelsMap().Sort()
					}
				case pdata.MetricDataTypeSum:
					for l := 0; l < m.Sum().DataPoints().Len(); l++ {
						m.Sum().DataPoints().At(l).LabelsMap().Sort()
					}
				case pdata.MetricDataTypeHistogram:
					for l := 0; l < m.Histogram().DataPoints().Len(); l++ {
						sortBuckets(m.Histogram().DataPoints().At(l))
						m.Histogram().DataPoints().At(l).LabelsMap().Sort()
					}
				case pdata.MetricDataTypeSummary:
					for l := 0; l < m.Summary().DataPoints().Len(); l++ {
						m.Summary().DataPoints().At(l).LabelsMap().Sort()
						m.Summary().DataPoints().At(l).QuantileValues().Sort(func(i, j int) bool {
							left := m.Summary().DataPoints().At(l).QuantileValues().At(i)
							right := m.Summary().DataPoints().At(l).QuantileValues().At(j)
							return left.Quantile() < right.Quantile()
						})
					}
				default:
					panic(fmt.Sprintf("unsupported metric data type %d", m.DataType()))
				}
			}
			il.Metrics().Sort(func(i, j int) bool {
				return il.Metrics().At(i).Name() < il.Metrics().At(j).Name()
			})
		}
		r.InstrumentationLibraryMetrics().Sort(func(i, j int) bool {
			left := r.InstrumentationLibraryMetrics().At(i).InstrumentationLibrary()
			right := r.InstrumentationLibraryMetrics().At(j).InstrumentationLibrary()
			if left.Name() == right.Name() {
				return left.Version() < right.Version()
			}
			return left.Name() < right.Name()
		})
		r.Resource().Attributes().Sort()
	}
}

func sortBuckets(hdp pdata.HistogramDataPoint) {
	buckets := make(sortableBuckets, len(hdp.ExplicitBounds()))
	for i := range hdp.ExplicitBounds() {
		buckets[i] = sortableBucket{hdp.BucketCounts()[i], hdp.ExplicitBounds()[i]}
	}
	sort.Sort(buckets)
	for i, bucket := range buckets {
		hdp.BucketCounts()[i], hdp.ExplicitBounds()[i] = bucket.count, bucket.bound
	}
}

type sortableBucket struct {
	count uint64
	bound float64
}

type sortableBuckets []sortableBucket

func (s sortableBuckets) Len() int {
	return len(s)
}

func (s sortableBuckets) Less(i, j int) bool {
	return s[i].bound < s[j].bound
}

func (s sortableBuckets) Swap(i, j int) {
	s[i].count, s[j].count = s[j].count, s[i].count
	s[i].bound, s[j].bound = s[j].bound, s[i].bound
}
```

File: cmd/pdatagen/internal/base_slices.go
Comment (bogdandrutu) [inline cmd/pdatagen/internal/base_slices.go:242]: I want to see an example of how to write this function. Do we have everything else that is needed?

Comment (jacobmarble) [reply cmd/pdatagen/internal/base_slices.go]: It looks like this comment is misleading, and even the implementation could be improved. Please take a look at the latest commit (forgive the rebase, there was a conflict).

The improvement is to make the less function reusable by passing it two `pdata` elements, rather than two indices.

Improved example code:
```golang
		lessFunc := func(a, b pdata.Metric) bool {
			return a.Name() < b.Name()
		}
		assert.EqualValues(t, expected.Metrics().Sort(lessFunc), got.Metrics().Sort(lessFunc))
		...
		r.InstrumentationLibraryMetrics().Sort(func(a, b pdata.InstrumentationLibraryMetrics) bool {
			if a.InstrumentationLibrary().Name() == b.InstrumentationLibrary().Name() {
				return a.InstrumentationLibrary().Version() < b.InstrumentationLibrary().Version()
			}
			return a.InstrumentationLibrary().Name() < b.InstrumentationLibrary().Name()
		})
```

**Comment by punya** [comment]
Sorry for the late feedback, but given that this is solely for tests, is it worth writing a single assertion that's implemented using reflection, rather than generating all this code? I was imagining a signature like

```go
// slice1 and slice2 are []T
// compareFunc is func(a, b T) int
func AssertEqualSlices(t *testing.T, slice1, slice2, compareFunc interface{}) { ... }
```

Clearly this signature is less self-documenting than the generated signatures. But could it be a better trade-off than all the code generation.

(This is also a matter of personal preference, so I would defer to the author or maintainers if they believe the original idea is a better fit.)

**Comment by jacobmarble** [comment]
@punya your comment reflects my comment about `StringMap.Sort()` and `AttributeMap.Sort()`. If `Sort()` is a method on those map/structs, then it should also exist on slice/structs, following the same pattern.

**Comment by punya** [comment]
@jacobmarble understood. I would make a different trade-off for those too, but I recognize that this is a personal preference.

**Comment by bogdandrutu** [comment]
@punya there are some metrics protocols that require ordering of timeseries/points. In that case I think we may need to do a sorting of the points by timestamp, so sorting is most likely to be needed for non-test code as well. Same may apply for logs or even span.

**Comment by jacobmarble** [comment]
@bogdandrutu is there anything left for me on this one?

File: cmd/pdatagen/internal/base_slices.go
Comment (bogdandrutu) [reply cmd/pdatagen/internal/base_slices.go]: @tigrannajaryan this is a bit different than golang, are we ok with this? Personally I like it.

Comment (tigrannajaryan) [reply cmd/pdatagen/internal/base_slices.go]: Yeah, I like it. Fairly standard sort func (even if not what golang's stdlib uses) and easy to understand.

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**
